aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -484,6 +497,24 @@ void replace_(const ITensorListRef functional_tensor, ITensorListRef other) {
   }
 }
 
+void propagate_xla_data(const Tensor& functional_tensor, const Tensor& other) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(isFunctionalTensor(functional_tensor));
+  if (functional_tensor.key_set().has(c10::DispatchKey::XLA)) {
+    at::_propagate_xla_data(at::functionalization::impl::unsafeGetFunctionalWrapper(functional_tensor)
+        ->value(), other);
+  }
+}
+
+void propagate_xla_data(const ITensorListRef functional_tensor, ITensorListRef other) {
```

Comment:
So it looks like the only reason this code lives in `FunctionalTensorWrapper.cpp` is because you need to handle the case where the input is either a tensor, or a list of tensors.

I think I'd prefer for these two functions to live directly in `ATen/ops/_propagate_xla_data.h` / `_propagate_xla_data.cpp`. Mostly just for locality: we have one xla-specific bit of functionalization, but it would be nice for it to all be localized to one file (I think it's fine to put it in that file even if you're putting it under the `at::functionalization::impl` namespace, since that makes it even clearer that the files just exists for functionalization).

Reply:
Sure, will make the corresponding change.

Reply:
Are you suggesting that we should gather all these helpers into `aten/src/ATen/native/_propagate_xla_data.*`? I'm not sure how we can have hand written code in `ATen/ops/_propagate_xla_data.*` since those are code-gened?

Reply:
oh hm... I would have expected a linker error, since the codegen should spit out a declaration for a new function `at::native::_propagate_xla_data`, that has no corresponding definition (do you see a codegen'd file locally at `build/aten/src/ATen/ops/_propagate_xla_data_native.h` that has that declaration? And originally I thought that you could stuff all the other relevant files there.

If you do end up needing to create a new cpp file to put the definition for that function in, I would suggest putting these implementation there too. If it isn't necessary to appease the build system, then leaving them here seems fine (can you add a comment above mentioning that this logic is XLA-specific and is a no-op for the normal functionalization flow?

Reply:
edit - nvm, I see you put the impl in `Copy.cpp`. You can just leave these implementations here (and add a quick comment)

Reply:
Adds comment to the header.